src/mk.rs

```diff
@@ -0,0 +1,143 @@
+    root_hash: Hash,
+}
+
+impl MerkleTree {
+    fn create<T: DataToHash>(data: Vec<T>) -> Option<Self> {
+        if data.is_empty() {
+            return None;
+        }
+        let leaves = MerkleTree::get_leaves_from(data);
+
+        let tree = MerkleTree::create_tree(leaves.clone());
+        let root_hash = tree.borrow().value;
+
+        Some(Self {
+            tree,
+            root_hash,
+            leaves,
+        })
+    }
+
+    fn get_leaves_from<T: DataToHash>(data: Vec<T>) -> Vec<MKNode> {
+        data.iter()
+            .map(|el| Node::new(get_hash_from_data(el), None, None, None))
+            .collect()
+    }
+
+    fn create_tree(mut leaves: Vec<MKNode>) -> MKNode {
+        while leaves.len() != 1 {
+            leaves = leaves
+                .chunks(2)
+                .map(|el| match el {
+                    [a, b] => {
+                        let node = Node::new(
+                            get_combined_hash(a.borrow().value, b.borrow().value),
+                            Some(vec![Rc::clone(a), Rc::clone(b)]),
+                            None,
+                            None,
+                        );
+                        Node::set_parent_and_siblings(&node, &[a, b]);
+                        return node;
+                    }
+                    // hash with itself
+                    [a] => {
+                        let node = Node::new(
+                            get_combined_hash(a.borrow().value, a.borrow().value),
+                            Some(vec![Rc::clone(a)]),
+                            None,
+                            None,
+                        );
+                        // does not have siblings
+                        Node::set_parent(&node, a);
+                        return node;
+                    }
+                    _ => panic!("unexpected chunk size"),
+                })
+                .collect();
+        }
+
+        return leaves.get(0).unwrap().to_owned();
+    }
+
+    pub fn get_leaves(&self) -> Vec<MKNode> {
+        self.leaves.to_owned()
+    }
+
+    pub fn add_leaf<T: DataToHash>(&mut self, data: T) {
+        let hash = get_hash_from_data(data);
+        let node = Node::new(hash, None, None, None);
+        self.leaves.push(node);
+        self.rebuild_tree();
+    }
+
+    pub fn delete_leaf<T: DataToHash>(&mut self, index: usize) {
+        if let Some(_) = self.leaves.get(index) {
+            self.leaves.remove(index);
+            self.rebuild_tree();
+        }
+    }
+
+    pub fn update_left<T: DataToHash>(&mut self, index: usize, data: T) {
+        if let Some(node) = self.leaves.get(index) {
+            node.borrow_mut().value = get_hash_from_data(data);
+            self.rebuild_tree();
+        }
+    }
+
+    fn rebuild_tree(&mut self) {
+        let tree = MerkleTree::create_tree(self.leaves.clone());
+        let root_hash = tree.borrow().value;
+        self.tree = tree;
+        self.root_hash = root_hash;
+    }
```

Comment:
If you are rebuilding the tree from scratch with every modification to the data, wouldn't it be the same to just ask for the data, modify it, and create a new Merkle Tree instance with the new data?

Since you went through the trouble of making each node modifiable with RefCell and Rc, you could try to only modify the nodes that need modifying (the ones that lead from the leaf to the root node).

Reply:
Otherwise, it would be better to just have the tree be inmutable.